sql/src/main/codegen/config.fmpp

```diff
@@ -0,0 +1,433 @@
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to you under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# This file is an FMPP (http://fmpp.sourceforge.net/) configuration file to
+# allow clients to extend Calcite's SQL parser to support application specific
+# SQL statements, literals or data types.
+#
+# Calcite's parser grammar file (Parser.jj) is written in javacc
+# (http://javacc.java.net/) with Freemarker (http://freemarker.org/) variables
+# to allow clients to:
+#   1. have custom parser implementation class and package name.
+#   2. insert new parser method implementations written in javacc to parse
+#      custom:
+#      a) SQL statements.
+#      b) literals.
+#      c) data types.
+#   3. add new keywords to support custom SQL constructs added as part of (2).
+#   4. add import statements needed by inserted custom parser implementations.
+#
+# Parser template file (Parser.jj) along with this file are packaged as
+# part of the calcite-core-<version>.jar under "codegen" directory.
+
+data: {
+  parser: {
+    # Generated parser implementation package and class name.
+    package: "org.apache.druid.sql.calcite.parser",
+    class: "DruidSqlParserImpl",
+
+    # List of additional classes and packages to import.
+    # Example. "org.apache.calcite.sql.*", "java.util.List".
+    imports: [
+      "org.apache.calcite.sql.SqlNode"
+      "org.apache.calcite.sql.SqlInsert"
+      "org.apache.druid.sql.calcite.parser.DruidSqlInsert"
+    ]
+
+    # List of new keywords. Example: "DATABASES", "TABLES". If the keyword is not a reserved
+    # keyword add it to 'nonReservedKeywords' section.
+    keywords: [
```

Comment:
Make it non reserved ?

Reply:
It would be ideal that "CLUSTER" is non-reserved since it won't break people's query's with CLUSTER/cluster being an identifier (datasource name, column name, etc). 
But on testing it out, the test case `INSERT INTO dst SELECT * FROM foo CLUSTER BY dim1` breaks. This is because the parser is assuming it is written like `INSERT INTO dst SELECT * FROM foo AS CLUSTER ...`, since CLUSTER can be used to name stuff. Reverting the change for now.

Reply:
Got it. In that case let's keep it as a reserved keyword.

Reply:
One thing that @abhishekagarwal87 suggested is to have a flag (in PlannerContext or PlannerConfig) that can be used to switch between the custom and the default parser. That would make sense in case we don't want existing queries with `cluster` identifiers to fail.

Reply:
IMO, we should always use our own parser. It gives us more control over our dialect, which we'll find useful for other things in the future.

I'm not sure if parser flags are a good idea or not, but if we do have them, IMO the best approach is to have "versions" of the entire parser. i.e. the flag would be `druid.sql.parser.version=2` or `druid.sql.parser.version=latest` rather than controlling specific features.

But, again, I'm not sure if we need this or not.

New keywords are added from time to time in other databases' SQL dialects. How do they typically handle this?

Reply:
Having two parsers creates great usability and testing issues: All tests would have to be run in both. Users would need a per-query way to switch parsers, which means the poor user has to understand which is in each one, which is a hard task even for developers. Automated tools that generate Druid SQL would not know what to do.

So, better is to get the solution right within a single parser.

Reply:
The parser flag is supposed to be at a druid cluster level. It's definitely not great but a workaround if we don't want to use a new reserved keyword. But if we do,  using `CLUSTERED BY` instead of `CLUSTER BY` will definitely make the chance of affecting queries very low so +1 to that.

Reply:
Yep, I might have been unclear in the original comment. Its general usage is not to be overridden per query but acts as a fallback in case there is some constraint due to which these changes would break existing queries. But going through the  `PARTITIONED BY`, `CLUSTERED BY` route, it makes sense to not consider that case since maintaining and deprecating the older parser would be unnecessary overhead.